tornado/ioloop.py

```diff
@@ -625,6 +632,35 @@ def add_future(self, future, callback):
         future.add_done_callback(
             lambda future: self.add_callback(callback, future))
 
+    def run_in_executor(self, executor, func, *args):
+        """Runs a function in a ``concurrent.futures.Executor``. If
+        ``executor`` is ``None``, the IO loop's default executor will be used.
+
+        Use `functools.partial` to pass keyword arguments to `func`.
+
+        """
+        if not _concurrent_futures_found:
+            raise RuntimeError(
+                "concurrent.futures is required to use IOLoop.run_in_executor")
+
+        if executor is None:
+            if not hasattr(self, '_executor'):
+                self._executor = ThreadPoolExecutor(max_workers=(cpu_count() * 5))
+            executor = self._executor
+
+        future = executor.submit(func, *args)
+        self.add_future(future, lambda future: future.result())
+        if sys.version_info >= (3, 5):
```

Comment:
Use the same pattern unconditionally. It's possible to use async/await on older python versions via cython. 

But `chain_future` isn't safe for this cross-thread use case. It needs to be something like `self.add_future(future, lambda f: chain_future(future, future_))`. There should be a test that a callback added to a future returned by `run_in_executor` resolves on the `IOLoop` thread.